AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,877 @@ class AzureIoTHub {
             }
+            _log("Connecting...");
+
+            local devPath = "/" + _connStrParsed.DeviceId;
+            local username = format("%s%s/api-version=%s", _connStrParsed.HostName, devPath, AZURE_API_VERSION);
+            local resourcePath = format("/devices%s/api-version=%s", devPath, AZURE_API_VERSION);
+            local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+            local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
+            local sas = AzureIoTHub.SharedAccessSignature(resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
+
+            local options = {
+                "username" : username,
+                "password" : sas
+            };
+
+            local url = "ssl://" + _connStrParsed.HostName;
+
+            _isConnecting = true;
+
+            _mqttclient.connect(url, _connStrParsed.DeviceId, options);
         }
 
-        // done is called everytime a message is available, params - error, data
-        function receive(done) {
-            _handlers.onEvent <- done;
+        // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
+        //
+        // Returns:                         Nothing.
+        function disconnect() {
+            if ((!_isDisconnected || _isConnecting) && !_isDisconnecting) {
+                _isDisconnecting = true;
+                _mqttclient.disconnect(_onDisconnect.bindenv(this));
+            }
+        }
 
-            if (done == null) {
-                _receivers.event.close();
-                // clear out handlers
-                return this;
+        // Checks if the client is connected to Azure IoT Hub.
+        //
+        // Returns:                         Boolean: true if the client is connected, false otherwise.
+        function isConnected() {
+            return _isConnected;
+        }
+
+        // Sends a message to Azure IoT Hub (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages).
+        //
+        // Parameters:
+        //     msg : AzureIoTHub.Message    Message to send.
+        //     onSend : Function            Callback called when the operation is completed or an error happens.
+        //          (optional)              The callback signature:
+        //                                  onSend(msg, error), where
+        //                                      msg :               The original message passed to the sendMessage() method.
+        //                                          AzureIoTHub.Message
+        //                                      error : Integer     0 if the operation is completed successfully, an error code otherwise.
+        //
+        // Returns:                         Nothing.
+        function sendMessage(msg, onSend = null) {
+            local tooManyRequests = _msgPendingQueue.len() >= _options.maxPendingSendRequests;
```

Comment:
This can go down to line 777 as if `!_isConnected` there is no need to have the variable.

Reply:
Done